File: src/gateway/collectors/ta_collector.py
"""TA collection adapter.

Calls TAOrchestrator.analyze() for each symbol and passes through
the full multi-timeframe result.  The Gateway does NOT dictate
timeframes, lookback periods, or any TA-specific configuration.
The TA engine owns all of that via TAConfig.

The gateway does NOT own a symbol list. Symbols are provided by the
caller (dashboard/API) at runtime for every cycle.
"""

from __future__ import annotations

import asyncio
import time
from datetime import UTC, datetime
from typing import Optional

from engine.shared.logging import get_logger
from engine.ta.orchestrator import TAOrchestrator
from gateway.config import GatewayConfig
from gateway.constants import PipelineStage
from gateway.context.models import TAResult, TASymbolResult
from gateway.observability.metrics import (
    GATEWAY_STAGE_ERRORS,
    GATEWAY_TA_CANDIDATES_PER_CYCLE,
    GATEWAY_TA_COLLECT_DURATION,
)

logger = get_logger(__name__)


class TACollector:
    """Collects TA analysis results for caller-provided symbols.

    The collector is a thin adapter between the Gateway pipeline and
    the TA engine.  It triggers analysis and maps the result into
    the Gateway's TASymbolResult model.  It does NOT configure the
    TA engine -- timeframes, lookback, and detection logic are
    entirely owned by the TA engine.
    """

    def __init__(
        self,
        *,
        ta_orchestrator: TAOrchestrator,
        config: GatewayConfig,
    ) -> None:
        self._orchestrator = ta_orchestrator
        self._config = config

    async def collect(
        self,
        *,
        symbols: list[str],
        trace_id: Optional[str] = None,
    ) -> TAResult:
        """Run TA analysis for the given symbols with bounded concurrency.

        Args:
            symbols: Symbols to analyse. Provided by the caller
                     (dashboard/API) -- never hardcoded.
            trace_id: Distributed trace ID for correlation.
        """
        if not symbols:
            logger.warning(
                "ta_collect_called_with_empty_symbols",
                extra={"trace_id": trace_id},
            )
            return TAResult(
                symbol_results=[],
                collected_at=datetime.now(UTC),
                duration_ms=0.0,
            )

        start = time.monotonic()
        semaphore = asyncio.Semaphore(self._config.max_concurrent_symbols)

        async def _analyze_symbol(symbol: str) -> TASymbolResult:
            async with semaphore:
                return await self._analyze_single(symbol, trace_id=trace_id)

        tasks = [_analyze_symbol(s) for s in symbols]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        symbol_results: list[TASymbolResult] = []
        for i, result in enumerate(results):
            if isinstance(result, Exception):
                symbol = symbols[i]
                logger.error(
                    "ta_symbol_analysis_exception",
                    extra={
                        "symbol": symbol,
                        "error": str(result),
                        "trace_id": trace_id,
                    },
                )
                GATEWAY_STAGE_ERRORS.labels(
                    stage=PipelineStage.TA_COLLECTOR,
                    error_type=type(result).__name__,
                ).inc()
                symbol_results.append(TASymbolResult(
                    symbol=symbol,
                    status="error",
                    error=str(result),
                ))
            else:
                symbol_results.append(result)

        elapsed_ms = (time.monotonic() - start) * 1000

        logger.info(
            "ta_collection_completed",
            extra={
                "symbols_requested": symbols,
                "symbols_total": len(symbols),
                "symbols_success": sum(1 for r in symbol_results if r.status == "success"),
                "duration_ms": round(elapsed_ms, 1),
                "trace_id": trace_id,
            },
        )

        return TAResult(
            symbol_results=symbol_results,
            collected_at=datetime.now(UTC),
            duration_ms=elapsed_ms,
        )

    async def _analyze_single(
        self,
        symbol: str,
        *,
        trace_id: Optional[str] = None,
    ) -> TASymbolResult:
        """Analyze a single symbol via the TA orchestrator.

        The orchestrator owns all TA logic: timeframe selection,
        candle fetching, pattern detection, snapshot building,
        alignment, and persistence.  This method simply maps
        the orchestrator's result dict into a TASymbolResult.
        """
        sym_start = time.monotonic()

        try:
            result = await self._orchestrator.analyze(symbol=symbol)

            sym_elapsed = time.monotonic() - sym_start
            GATEWAY_TA_COLLECT_DURATION.labels(symbol=symbol).observe(sym_elapsed)

            status = result.get("status", "error")

            smc_candidates = result.get("smc_candidates", [])
            snd_candidates = result.get("snd_candidates", [])

            if status == "success":
                GATEWAY_TA_CANDIDATES_PER_CYCLE.labels(framework="smc").observe(
                    result.get("smc_candidates_count", 0),
                )
                GATEWAY_TA_CANDIDATES_PER_CYCLE.labels(framework="snd").observe(
                    result.get("snd_candidates_count", 0),
                )

            return TASymbolResult(
                symbol=symbol,
                htf_timeframes=result.get("htf_timeframes", []),
                ltf_timeframes=result.get("ltf_timeframes", []),
                status=status,
                smc_candidates=smc_candidates,
                snd_candidates=snd_candidates,
                snapshots=result.get("snapshots", {}),
                alignment=result.get("alignment", {}),
                overall_trend=result.get("overall_trend", "NEUTRAL"),
                error=result.get("error"),
            )

        except Exception as exc:
            sym_elapsed = time.monotonic() - sym_start
            GATEWAY_TA_COLLECT_DURATION.labels(symbol=symbol).observe(sym_elapsed)
            GATEWAY_STAGE_ERRORS.labels(
                stage=PipelineStage.TA_COLLECTOR,
                error_type=type(exc).__name__,
            ).inc()

            logger.error(
                "ta_single_symbol_failed",
                extra={
                    "symbol": symbol,
                    "error": str(exc),
                    "trace_id": trace_id,
                },
                exc_info=True,
            )

            return TASymbolResult(
                symbol=symbol,
                status="error",
                error=str(exc),
            )

File: src/gateway/internal/collectors/ta_collector.go
package collectors

import (
	"context"
	"encoding/json"
	"sort"
	"strings"
	"time"

	"github.com/rs/zerolog"

	"github.com/flamegreat-1/etradie/src/auth"
	"github.com/flamegreat-1/etradie/src/gateway/internal/config"
	"github.com/flamegreat-1/etradie/src/gateway/internal/constants"
	"github.com/flamegreat-1/etradie/src/gateway/internal/infra"
	"github.com/flamegreat-1/etradie/src/gateway/internal/models"
	"github.com/flamegreat-1/etradie/src/gateway/internal/observability"
)

// TACollector calls the Python TA engine via HTTP for each symbol
// with bounded concurrency, then maps results into gateway models.
// Results are cached in Redis to avoid redundant HTTP calls within
// the configured TTL window.
type TACollector struct {
	engine   *infra.EngineHTTPClient
	redis    *infra.RedisClient
	cfg      *config.Config
	cacheTTL time.Duration
	log      zerolog.Logger
}

// NewTACollector creates a TACollector backed by the engine HTTP client
// with optional Redis caching. If redis is nil or TTL is 0, caching is disabled.
func NewTACollector(engine *infra.EngineHTTPClient, redis *infra.RedisClient, cfg *config.Config) *TACollector {
	return &TACollector{
		engine:   engine,
		redis:    redis,
		cfg:      cfg,
		cacheTTL: time.Duration(cfg.TACacheTTLSeconds) * time.Second,
		log:      observability.Logger("ta_collector"),
	}
}

// taCacheKey builds a deterministic cache key from the user ID and sorted symbol list.
// The userID ensures cache isolation between tenants.
func taCacheKey(userID string, symbols []string) string {
	sorted := make([]string, len(symbols))
	copy(sorted, symbols)
	sort.Strings(sorted)
	if userID != "" {
		return userID + ":" + strings.Join(sorted, ",")
	}
	return strings.Join(sorted, ",")
}

// Collect runs TA analysis for the given symbols via a single HTTP call
// to the Python engine. The Python side processes them in parallel.
func (c *TACollector) Collect(ctx context.Context, symbols []string, traceID string, bypassCache bool) (*models.TAResult, error) {
	if len(symbols) == 0 {
		c.log.Warn().Str("trace_id", traceID).Msg("ta_collect_called_with_empty_symbols")
		return &models.TAResult{
			CollectedAt: time.Now().UTC(),
		}, nil
	}

	// Extract userID for cache key scoping (multi-tenant isolation).
	userID := auth.UserIDFromContext(ctx)

	// Check cache first (unless bypassing).
	if !bypassCache {
		if cached := c.getFromCache(ctx, userID, symbols, traceID); cached != nil {
			return cached, nil
		}
	}

	start := time.Now()

	// Call the Python engine once with all symbols.
	reqBody := map[string]interface{}{
		"symbols":  symbols,
		"trace_id": traceID,
	}

	resp, err := c.engine.PostJSON(ctx, "/internal/ta/analyze", reqBody)

	elapsed := time.Since(start)
	observability.GatewayTACollectDuration.Observe(elapsed.Seconds())

	if err != nil {
		observability.GatewayStageErrors.WithLabelValues(
			constants.StageTACollector.String(), "http_error",
		).Inc()
		c.log.Error().
			Err(err).
			Str("trace_id", traceID).
			Msg("ta_collection_http_failed")
		return nil, err
	}

	// Parse symbol_results from the response.
	results := c.parseSymbolResults(resp, symbols, traceID)

	elapsedMs := float64(elapsed.Milliseconds())
	successCount := 0
	for i := range results {
		if results[i].Status == "success" {
			successCount++
		}
	}

	c.log.Info().
		Strs("symbols_requested", symbols).
		Int("symbols_total", len(symbols)).
		Int("symbols_success", successCount).
		Float64("duration_ms", elapsedMs).
		Str("trace_id", traceID).
		Msg("ta_collection_completed")

	result := &models.TAResult{
		SymbolResults: results,
		CollectedAt:   time.Now().UTC(),
		DurationMs:    elapsedMs,
	}

	// Cache successful result.
	c.storeInCache(ctx, userID, symbols, result, traceID)

	return result, nil
}

func (c *TACollector) getFromCache(ctx context.Context, userID string, symbols []string, traceID string) *models.TAResult {
	if c.redis == nil || c.cacheTTL <= 0 {
		return nil
	}

	key := taCacheKey(userID, symbols)
	raw, err := c.redis.GetRaw(ctx, constants.GatewayCacheNamespace, constants.TAResultCacheKeyPrefix+":"+key)
	if err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("ta_cache_read_error")
		return nil
	}
	if raw == nil {
		return nil
	}

	// Single-pass unmarshal directly into the typed struct.
	var result models.TAResult
	if err := json.Unmarshal(raw, &result); err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("ta_cache_unmarshal_error")
		return nil
	}

	c.log.Info().
		Strs("symbols", symbols).
		Str("trace_id", traceID).
		Msg("ta_result_served_from_cache")
	return &result
}

func (c *TACollector) storeInCache(ctx context.Context, userID string, symbols []string, result *models.TAResult, traceID string) {
	if c.redis == nil || c.cacheTTL <= 0 {
		return
	}

	key := taCacheKey(userID, symbols)
	if err := c.redis.Set(ctx, constants.GatewayCacheNamespace, constants.TAResultCacheKeyPrefix+":"+key, result, c.cacheTTL); err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("ta_cache_write_error")
	}
}

func (c *TACollector) parseSymbolResults(
	resp map[string]interface{},
	symbols []string,
	traceID string,
) []models.TASymbolResult {
	rawResults, ok := resp["symbol_results"]
	if !ok {
		c.log.Warn().Str("trace_id", traceID).Msg("ta_response_missing_symbol_results")
		var results []models.TASymbolResult
		for _, sym := range symbols {
			results = append(results, models.TASymbolResult{
				Symbol:       sym,
				Status:       "error",
				Error:        "missing symbol_results in response",
				OverallTrend: "NEUTRAL",
			})
		}
		return results
	}

	resultSlice, ok := rawResults.([]interface{})
	if !ok {
		c.log.Warn().Str("trace_id", traceID).Msg("ta_response_symbol_results_not_array")
		return nil
	}

	var results []models.TASymbolResult
	for _, raw := range resultSlice {
		resultMap, ok := raw.(map[string]interface{})
		if !ok {
			continue
		}

		sr := models.TASymbolResult{
			Symbol:        getStringField(resultMap, "symbol"),
			Status:        getStringField(resultMap, "status"),
			OverallTrend:  getStringFieldDefault(resultMap, "overall_trend", "NEUTRAL"),
			Error:         getStringField(resultMap, "error"),
			HTFTimeframes: getStringSlice(resultMap, "htf_timeframes"),
			LTFTimeframes: getStringSlice(resultMap, "ltf_timeframes"),
		}

		sr.SMCCandidates = getMapSlice(resultMap, "smc_candidates")
		sr.SnDCandidates = getMapSlice(resultMap, "snd_candidates")
		sr.Snapshots = getNestedMapMap(resultMap, "snapshots")
		sr.Alignment = getNestedMapMap(resultMap, "alignment")

		if sr.Status == "success" {
			observability.GatewayTACandidatesPerCycle.WithLabelValues("smc").Observe(float64(len(sr.SMCCandidates)))
			observability.GatewayTACandidatesPerCycle.WithLabelValues("snd").Observe(float64(len(sr.SnDCandidates)))
		}

		results = append(results, sr)
	}

	return results
}

// JSON response field helpers.

func getStringField(m map[string]interface{}, key string) string {
	v, ok := m[key]
	if !ok || v == nil {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return ""
	}
	return s
}

func getStringFieldDefault(m map[string]interface{}, key, def string) string {
	s := getStringField(m, key)
	if s == "" {
		return def
	}
	return s
}

func getStringSlice(m map[string]interface{}, key string) []string {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	slice, ok := v.([]interface{})
	if !ok {
		return nil
	}
	out := make([]string, 0, len(slice))
	for _, item := range slice {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func getMapSlice(m map[string]interface{}, key string) []map[string]interface{} {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	slice, ok := v.([]interface{})
	if !ok {
		return nil
	}
	out := make([]map[string]interface{}, 0, len(slice))
	for _, item := range slice {
		if mp, ok := item.(map[string]interface{}); ok {
			out = append(out, mp)
		}
	}
	return out
}

func getNestedMapMap(m map[string]interface{}, key string) map[string]map[string]interface{} {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	outer, ok := v.(map[string]interface{})
	if !ok {
		return nil
	}
	out := make(map[string]map[string]interface{}, len(outer))
	for k, val := range outer {
		if inner, ok := val.(map[string]interface{}); ok {
			out[k] = inner
		}
	}
	return out
}
